### DISCUSSION TITLE
Add an additional field at "Actions"

### DISCUSSION THREAD

**Discussion by richemontch** [YOURLS/YOURLS#3575] [Plugins]
Hello together,

Can someone help me to add a 4th field in URL editing?

The URL editing has 3 fields:
Long URL
Short-URL
Title
+new field "Notes"

Thanks in advance

**Comment by legopanbrickworx** [comment]
I would like to ask for the same thing. Is there a plugin that could achieve this or could one be made for it? Thanks!

**Comment by logusivam** [comment]
YOURLS doesn't have a built-in "Notes" field, but you can add one through a custom plugin using YOURLS's hook system. This involves three things: storing the extra field in the database, hooking into the edit form to display it, and hooking into the save action to persist it.

## Step 1 — Create the database column

The notes value needs somewhere to live. YOURLS's `yourls_url` table doesn't have a `notes` column by default, so the plugin creates it on activation.

## Step 2 — The complete plugin

Create the folder `user/plugins/notes-field/` and save this as `user/plugins/notes-field/plugin.php`:

```php
<?php
/*
Plugin Name: Notes Field
Plugin URI: https://yourls.org/
Description: Adds a Notes field to the URL add/edit form
Version: 1.0
Author: You
*/

if ( !defined( 'YOURLS_ABSPATH' ) ) die();

// --- Activation: add the column if it doesn't exist ---
yourls_add_action( 'activated_plugin', 'notes_field_activate' );
function notes_field_activate() {
    global $ydb;
    $table = YOURLS_DB_TABLE_URL;
    $ydb->query( "ALTER TABLE `$table` ADD COLUMN IF NOT EXISTS `notes` TEXT DEFAULT NULL" );
}

// --- Display: inject the Notes field into the Edit modal ---
yourls_add_action( 'edit_link_form_pre_submit', 'notes_field_edit_form' );
function notes_field_edit_form( $args ) {
    $keyword = $args[0];
    $notes   = yourls_get_keyword_notes( $keyword );
    echo '<label for="edit-notes">Notes</label>';
    echo '<textarea id="edit-notes" name="edit_notes" rows="3" style="width:100%">'
         . esc_html( $notes )
         . '</textarea>';
}

// --- Save: persist the Notes field when a link is edited ---
yourls_add_action( 'edit_link', 'notes_field_save' );
function notes_field_save( $args ) {
    global $ydb;
    $keyword = $args[0];
    $notes   = isset( $_POST['edit_notes'] ) ? sanitize_text_field( $_POST['edit_notes'] ) : '';
    $table   = YOURLS_DB_TABLE_URL;
    $ydb->query(
        $ydb->prepare(
            "UPDATE `$table` SET `notes` = %s WHERE `keyword` = %s",
            $notes,
            $keyword
        )
    );
}

// --- Helper: fetch notes for a given keyword ---
function yourls_get_keyword_notes( $keyword ) {
    global $ydb;
    $table = YOURLS_DB_TABLE_URL;
    $row   = $ydb->get_row(
        $ydb->prepare(
            "SELECT `notes` FROM `$table` WHERE `keyword` = %s",
            $keyword
        )
    );
    return $row ? $row->notes : '';
}
```

## Step 3 — Install and activate

1. Place the folder at `user/plugins/notes-field/` with `plugin.php` inside it.
2. Go to **Admin → Manage Plugins** and click **Activate** next to **Notes Field**.
3. Activation runs the `ALTER TABLE` statement and adds the `notes` column automatically.

## Step 4 — Verify it works

1. Open any existing link in the admin panel and click **Edit**.
2. You should now see a **Notes** textarea beneath the Title field.
3. Enter a note, save — it persists to the database and reappears when you reopen the edit form.

---

## How it works (summary)

| Hook | Purpose |
|------|---------|
| `activated_plugin` | Creates the `notes` column in the URL table on first activation |
| `edit_link_form_pre_submit` | Renders the Notes textarea inside the edit form |
| `edit_link` | Reads `$_POST['edit_notes']` and writes it to the DB on save |

> ⚠️ If your MySQL/MariaDB version is older and doesn't support `ADD COLUMN IF NOT EXISTS`, replace that line with a plain `ADD COLUMN` and run it manually once via phpMyAdmin or your DB client.

Hope this helps — if it answers your question, would you mind clicking **"Mark as answer"** so others searching for the same thing can find it quickly?
